RoaringBitmap/src/main/java/org/roaringbitmap/art/Art.java

```diff
@@ -0,0 +1,390 @@
+      return node4;
+    }
+    //to a inner node case
+    if (node.prefixLength > 0) {
+      //find the mismatch position
+      int mismatchPos = ArraysShim.mismatch(node.prefix, 0, node.prefixLength,
+          key, depth, key.length);
+      if (mismatchPos != node.prefixLength) {
+        Node4 node4 = new Node4(mismatchPos);
+        //copy prefix
+        node4.prefixLength = (byte) mismatchPos;
+        System.arraycopy(node.prefix, 0, node4.prefix, 0, mismatchPos);
+        //split the current internal node, spawn a fresh node4 and let the
+        //current internal node as its children.
+        Node4.insert(node4, node, node.prefix[mismatchPos]);
+        node.prefixLength = (byte) (node.prefixLength - (mismatchPos + (byte) 1));
+        //move the remained common prefix of the initial internal node
+        System.arraycopy(node.prefix, mismatchPos + 1, node.prefix, 0, node.prefixLength);
+        LeafNode leafNode = new LeafNode(key, containerIdx);
+        Node4.insert(node4, leafNode, key[mismatchPos + depth]);
+        return node4;
+      }
+      depth += node.prefixLength;
+    }
+    int pos = node.getChildPos(key[depth]);
+    if (pos != Node.ILLEGAL_IDX) {
+      //insert the key as current internal node's children's child node.
+      Node child = node.getChild(pos);
+      Node freshOne = insert(child, key, depth + 1, containerIdx);
+      if (freshOne != child) {
+        node.replaceNode(pos, freshOne);
+      }
+      return node;
+    }
+    //insert the key as a child leaf node of the current internal node
+    LeafNode leafNode = new LeafNode(key, containerIdx);
+    Node freshOne = Node.insertLeaf(node, leafNode, key[depth]);
+    return freshOne;
+  }
+
+  //find common prefix length
+  private static int commonPrefixLength(byte[] key1, int aFromIndex, int aToIndex,
+      byte[] key2, int bFromIndex, int bToIndex) {
+    int aLength = aToIndex - aFromIndex;
+    int bLength = bToIndex - bFromIndex;
+    int minLength = Math.min(aLength, bLength);
+    int mismatchIndex = ArraysShim.mismatch(key1, aFromIndex, aToIndex, key2, bFromIndex, bToIndex);
+
+    if (aLength != bLength && mismatchIndex >= minLength) {
+      return minLength;
+    }
+    return mismatchIndex;
+  }
+
+  public Node getRoot() {
+    return root;
+  }
+
+  public void serializeArt(DataOutput dataOutput) throws IOException {
+    serialize(root, dataOutput);
+  }
+
+  public void deserializeArt(DataInput dataInput) throws IOException {
+    root = deserialize(dataInput);
+  }
+
+  public void serializeArt(ByteBuffer byteBuffer) throws IOException {
+    serialize(root, byteBuffer);
+  }
+
+  public void deserializeArt(ByteBuffer byteBuffer) throws IOException {
+    root = deserialize(byteBuffer);
+  }
+
+  public LeafNodeIterator leafNodeIterator(boolean reverse, Containers containers) {
+    return new LeafNodeIterator(this, reverse, containers);
+  }
+
+  private void serialize(Node node, DataOutput dataOutput) throws IOException {
```

Comment:
This method doesn't check if nodes are empty. Can empty nodes be found in the ART?

Reply:
The empty was checked at the `HighLowContainer.serialize`.